### ISSUE TITLE
`copy.deepcopy` does not copy gradients of nn.Parameter

### ISSUE THREAD

**Issue by soulitzer** [pytorch/pytorch#95711]
Deep copy now copies the .grad field for tensors, but it does not work for Parameter

```
import torch
import copy

c = torch.nn.Linear(1, 10)

c.weight.grad = torch.rand((10, 1))

print(c.weight.grad)
d = copy.deepcopy(c)
print(d.weight.grad)
```
Tested with version 1.13.1. Running the code from @albanD gives the expected result. It would be great if we could find a way to copy an entire (larger) model including its gradients without having to copy every one by hand.

_Originally posted by @jhuebotter in https://github.com/pytorch/pytorch/issues/3307#issuecomment-1425753484_

**Comment by soulitzer** [comment]
Seems like an easy fix if we do want to do this. Just add the following to nn.Parameter's  `__deepcopy__`
```
if self.grad is not None:
    result.grad = self.grad.__deepcopy__(memo)
```

**Comment by albanD** [comment]
Ho yea we should fix that.
There are a couple of issues in Parameter deepcopy cc @jbschlosser I think you still have a PR in flight about this that Richard was helping with?

**Comment by jbschlosser** [comment]
You're referring to #91636, yes?

**Comment by albanD** [comment]
Just after branch cut is the best time to land BC-breaking stuff haha

**Comment by gchanan** [comment]
@zou3519 should be back in a ~week -- do you want to take this on now @jbschlosser or wait?

**Comment by jbschlosser** [comment]
Note that the fix in #91636 doesn't solve this issue, even though it dispatches to `Tensor.__deepcopy__()`, which correctly copies the grad. We lose the grad after the call to `torch.Tensor._make_subclass()` in `Parameter.__new__()` since it `detach()`es.
